posit-dev/positron
# Fix issues around extension host restarts

### Intent

Addresses #2683. 

The problem here is pretty ugly. When the extension host is stopped in the middle of a session, it yanks out the middleware between the front end and the R/Python kernel. So the UI and runtime are still alive, but all the plumbing needed for them to talk to each other is gone.

### Approach

The main idea here is to reuse all the codepaths around reconnecting to a kernel after a front-end restart (which we _do_ have pretty good support for).

<img width="386" alt="image" src="https://github.com/posit-dev/positron/assets/470418/f8382bbe-c5b9-4cd5-9376-7a84100be40d">

When the extension host is shut down, we mark all the runtimes as exited, but we also store them in a cache on the frontend. When the extension host comes back online, we use the cache to reconnect to the sessions. The reconnection logic takes care of all the restoration of the middleware.

### QA Notes

- There may be other issues around these scenarios; this fix only attempts to address the most obvious for Beta. 
- The "Save Workspace As" experience is still not great; it will disconnect your session but not reconnect it (it can't because the session is in the old workspace). Nothing should be unresponsive, though, and you should be able to start a new session and keep on going.
